File: src/app/src/main/java/com/tylermayoff/dynamicwallpaper/DynamicWallpaperService.java
package com.tylermayoff.dynamicwallpaper;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.os.Handler;
import android.service.wallpaper.WallpaperService;
import android.view.MotionEvent;
import android.view.SurfaceHolder;

import java.io.File;
import java.util.Calendar;

/*
public class DynamicWallpaperService extends WallpaperService {

    private DynamicWallpaperEngine engine;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            String action = intent.getAction();
            if (action == null) return super.onStartCommand(intent, flags, startId);

            if (action.equals("NEXT") && engine != null) {
                engine.NextImage();
            }
        }

        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public Engine onCreateEngine() {
        engine = new DynamicWallpaperEngine(this);
        return engine;
    }

    public class DynamicWallpaperEngine extends Engine {

        public ThemeConfig themeConfig;

        private final Handler handler = new Handler();
        private final Runnable drawRunner = () -> draw();

        private int scrHeight;
        private int scrWidth;
        private int currentIndex;
        private boolean visible = true;

        // Animation Rules
        private int FADE_MILLI = 500;
        private int FADE_STEP = 5;
        private int ALPHA_STEP = 255 / (FADE_MILLI / FADE_STEP);

        private Paint alphaPaint = new Paint();
        private int currentAlpha = 0;
        private boolean changingImage = true;

        private AlarmManager alarmManager;

        Context context;

        public DynamicWallpaperEngine(Context c) {
            context = c;

            SharedPreferences sharedPreferences = getSharedPreferences(context.getString(R.string.preferences_file_key), Context.MODE_PRIVATE);
            String activeTheme = sharedPreferences.getString(context.getString(R.string.preferences_active_theme), "");

            if (themeConfig == null && !activeTheme.isEmpty())
                themeConfig = new ThemeConfig(new File(getFilesDir() + "/theme/" + activeTheme));

            alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
            Intent intent = new Intent(context, DynamicWallpaperService.class);
            intent.putExtra("startId", 5);
            intent.setAction("NEXT");
            PendingIntent pendingIntent = PendingIntent.getService(context, 0, intent, 0);

            // Setup next background change
            currentIndex = themeConfig.GetLastTimeIndex();
            Calendar nextAlarm = themeConfig.GetNextTime(currentIndex);
            alarmManager.set(AlarmManager.RTC, nextAlarm.getTimeInMillis(), pendingIntent);

            changingImage = true;

            handler.post(drawRunner);
        }

        @Override
        public void onVisibilityChanged(boolean visible) {
            this.visible = visible;
            if(visible)
                handler.post(drawRunner);
            */
/*else
                handler.removeCallbacks(drawRunner);*//*

        }

        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            this.scrWidth = width;
            this.scrHeight = height;

            super.onSurfaceChanged(holder, format, width, height);
        }

        @Override
        public void onTouchEvent(MotionEvent event) {
            // Do nothing
        }

        private void draw() {

            SurfaceHolder holder = getSurfaceHolder();
            Canvas canvas = null;
            if (!changingImage) return;

            try {
                canvas = holder.lockCanvas();
                if (canvas != null) {
                    if (themeConfig.images.size() == 0)
                        return;

                    int originalWidth = themeConfig.images.get(currentIndex).getWidth();
                    int originalHeight = themeConfig.images.get(currentIndex).getHeight();

                    int scale = scrHeight / originalHeight * 2;

                    float xTranslation = (scrWidth - originalWidth * scale) / 2.0f;
                    float yTranslation = (scrHeight - originalHeight * scale) / 2.0f;

                    Matrix transformation = new Matrix();
                    transformation.postTranslate(xTranslation, yTranslation);
                    transformation.preScale(scale, scale);

                    alphaPaint.setAlpha(currentAlpha);
                    currentAlpha += ALPHA_STEP;
                    alphaPaint.setFilterBitmap(true);
                    canvas.drawBitmap(themeConfig.images.get(currentIndex), transformation, alphaPaint);

                    if (255 - currentAlpha < 10) {
                        // Finished drawing
                        currentAlpha = 0;
                        changingImage = false;

                        handler.removeCallbacks(drawRunner);
                    }
                }
            }
            finally {
                if (canvas != null) holder.unlockCanvasAndPost(canvas);
            }

            handler.removeCallbacks(drawRunner);
            if (visible && changingImage) {
                handler.postDelayed(drawRunner, FADE_STEP);
            }
        }

        public void NextImage () {
            currentIndex++;
            if (currentIndex >= themeConfig.images.size())
                currentIndex = 0;

            changingImage = true;

            // Setup next background change
            Intent intent = new Intent(context, DynamicWallpaperService.class);
            intent.setAction("NEXT");
            PendingIntent pendingIntent = PendingIntent.getService(context, 0, intent, 0);

            // Setup next background change
            Calendar nextAlarm = themeConfig.GetNextTime(currentIndex);
            alarmManager.set(AlarmManager.RTC, nextAlarm.getTimeInMillis(), pendingIntent);

            handler.post(drawRunner);
        }
    }
}
*/

File: app/src/main/java/com/tylermayoff/dynamicwallpaper/DynamicWallpaperService.kt
package com.tylermayoff.dynamicwallpaper

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.os.Handler
import android.os.Looper
import android.service.wallpaper.WallpaperService
import android.view.SurfaceHolder
import com.tylermayoff.dynamicwallpaper.util.AppSettings
import com.tylermayoff.dynamicwallpaper.util.ThemeConfiguration
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.*

class DynamicWallpaperService : WallpaperService() {

    lateinit var engine : DynamicWallpaperEngine

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent != null) {
            if (intent.action?.equals("NEXT_IMG")!!) {
                engine.nextImage()
            }
        }

        return super.onStartCommand(intent, flags, startId)
    }

    override fun onCreateEngine(): Engine {
        engine = DynamicWallpaperEngine(this as Context)
        return engine
    }

    inner class DynamicWallpaperEngine(private val context: Context) : WallpaperService.Engine() {

        private var sharedPreferences : SharedPreferences = getSharedPreferences(context.getString(R.string.preferences_file_key), Context.MODE_PRIVATE)
        private var appSettings = AppSettings.getInstance(context)

        private val show : Runnable = Runnable { show() }
        private val changeImage : Runnable = Runnable { changeImage() }
        private val handler : Handler = Handler(Looper.getMainLooper())
        private val alarmManager : AlarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager

        // Theme Specifics
        private var activeThemeName : String
        private var currentImageIndex : Int = 0

        // Screen specifics
        private var screenHeight : Int = 0
        private var screenWidth : Int = 0
        private var wallpaperIsVisible : Boolean = true

        // Draw
        private val fadeMilli : Long = 500
        private val fadeStep : Long = 5
        private val alphaStep : Long = 255 / (fadeMilli / fadeStep)

        var alphaPaint : Paint = Paint()
        var currentAlpha : Int = 0
        var doneAnimating : Boolean = false

        init {

            activeThemeName = sharedPreferences.getString(context.getString(R.string.preferences_active_theme), "")!!

            if (activeThemeName.isEmpty()) {
                // TODO Redirect to activity

            }

            // Setup alarm
            val intent = Intent(context, DynamicWallpaperEngine::class.java)
            intent.action = "NEXT_IMG"
            val pendingIntent : PendingIntent = PendingIntent.getService(context, 0, intent, 0)

            val nextAlarm = LocalDateTime.of(LocalDate.now(), appSettings.themeConfig!!.getNextChangeTime())
            alarmManager.set(AlarmManager.RTC, nextAlarm.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli(), pendingIntent)

            handler.post(changeImage)
        }

        override fun onSurfaceChanged(holder: SurfaceHolder?, format: Int, width: Int, height: Int) {
            this.screenHeight = height
            this.screenWidth = width

            super.onSurfaceChanged(holder, format, width, height)
        }

        override fun onVisibilityChanged(visible: Boolean) {
            if (visible){
                this.wallpaperIsVisible = visible
                doneAnimating = true
                handler.post(show)
            } else {
                handler.removeCallbacks(changeImage)
            }

            super.onVisibilityChanged(visible)
        }

        fun nextImage () {
            currentImageIndex++
            if (currentImageIndex >= appSettings.themeConfig!!.images.size)
                currentImageIndex = 0

            currentAlpha = 0
            doneAnimating = false

            val intent = Intent(context, DynamicWallpaperEngine::class.java)
            intent.action = "NEXT_IMG"
            val pendingIntent : PendingIntent = PendingIntent.getService(context, 0, intent, 0)

            val nextAlarm = LocalDateTime.of(LocalDate.now(), appSettings.themeConfig!!.getNextChangeTime())
            alarmManager.set(AlarmManager.RTC, nextAlarm.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli(), pendingIntent)
        }

        private fun show () {

            var canvas : Canvas? = null
            try {
                canvas = surfaceHolder.lockCanvas()
                if (canvas != null) {
                    val img = appSettings.themeConfig!!.getCurrentBitmap()
                    val originalHeight = img.height
                    val originalWidth = img.width

                    val scale : Float = screenHeight / originalHeight * 2.0f
                    val xTranslate = (screenWidth - originalWidth * scale) / 2.0f
                    val yTranslate = (screenHeight - originalHeight * scale) / 2.0f

                    val transformation = Matrix()
                    transformation.postTranslate(xTranslate, yTranslate)
                    transformation.preScale(scale, scale)

                    val p = Paint()
                    p.isFilterBitmap = true
                    canvas.drawBitmap(img, transformation, p)
                }
            } finally {
                if (canvas != null) surfaceHolder.unlockCanvasAndPost(canvas)
            }
        }

        private fun changeImage () {
            var canvas : Canvas? = null
            try {
                canvas = surfaceHolder.lockCanvas()
                if (canvas != null) {
                    val img = appSettings.themeConfig!!.getCurrentBitmap()
                    val originalHeight = img.height
                    val originalWidth = img.width

                    val scale : Float = screenHeight / originalHeight * 2.0f
                    val xTranslate = (screenWidth - originalWidth * scale) / 2.0f
                    val yTranslate = (screenHeight - originalHeight * scale) / 2.0f

                    val transformation = Matrix()
                    transformation.postTranslate(xTranslate, yTranslate)
                    transformation.preScale(scale, scale)

                    alphaPaint.alpha = currentAlpha
                    alphaPaint.isFilterBitmap = true
                    currentAlpha += alphaStep.toInt()

                    canvas.drawBitmap(img, transformation, alphaPaint)
                    // Done animating
                    if (255 - currentAlpha < 10)
                        doneAnimating = true

                }
            } finally {
                if (canvas != null) surfaceHolder.unlockCanvasAndPost(canvas)
            }

            handler.removeCallbacks(changeImage)
            if (wallpaperIsVisible && !doneAnimating) {
                handler.postDelayed(changeImage, fadeStep)
            }
        }
    }
}